### ISSUE TITLE
CA2257 should not be reported on nested types

### ISSUE THREAD

**Issue by jkoritzinsky** [dotnet/roslyn-analyzers#7106]
### Analyzer

**Diagnostic ID**: [CA2257](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2257): `Members defined on an interface with the 'DynamicInterfaceCastableImplementationAttribute' should be 'static'`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 8.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

The analyzer reports that nested types in a type with `[DynamicInterfaceCastableImplementation]` should be static, but the rationale for that rule does not apply to nested types.

### Steps To Reproduce

```csharp

interface Foo
{
   void Bar();
}

[DynamicInterfaceCastableImplementation]
interface FooImpl
{
     void Foo.Bar() {}

     private class C {}
}

```

### Expected behavior

No diagnostic

### Actual behavior

A diagnostic on the class declaration
